include/aws/common/condition_variable.h

```diff
@@ -0,0 +1,87 @@
+#ifndef AWS_CONDITION_VARIABLE_H
+#define AWS_CONDITION_VARIABLE_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <pthread.h>
+#endif
+
+struct aws_mutex;
+
+struct aws_condition_variable {
+#ifdef _WIN32
+    CONDITION_VARIABLE condition_handle;
+#else
+    pthread_cond_t condition_handle;
+#endif
+};
+
+/**
+ * Static initializer for condition variable.
+ * You can do something like struct aws_condition_variable var = AWS_CONDITION_VARIABLE_INIT;
+ */
+#ifdef _WIN32
+#define AWS_CONDITION_VARIABLE_INIT                                                                                    \
+{ .condition_handle = CONDITION_VARIABLE_INIT }
+#else
+#define AWS_CONDITION_VARIABLE_INIT                                                                                    \
+{ .condition_handle = PTHREAD_COND_INITIALIZER }
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Initializes a condition variable.
+ */
+AWS_COMMON_API int aws_condition_variable_init (struct aws_condition_variable *condition_variable);
+
+/**
+ * Cleans up a condition variable.
+ */
+AWS_COMMON_API void aws_condition_variable_clean_up (struct aws_condition_variable *condition_variable);
+
+/**
+ * Notifies/Wakes one waiting thread
+ */
+AWS_COMMON_API int aws_condition_variable_notify_one (struct aws_condition_variable *condition_variable);
+
+/**
+ * Notifies/Wakes all waiting threads.
+ */
+AWS_COMMON_API int aws_condition_variable_notify_all (struct aws_condition_variable *condition_variable);
+
+/**
+ * Waits the calling thread on a notification from another thread.
+ */
+AWS_COMMON_API int aws_condition_variable_wait (struct aws_condition_variable *condition_variable, struct aws_mutex *mutex);
```

Comment:
You need another parameter for a predicate to handle the spurious wakeup case.

Reply:
>When using condition variables there is always a Boolean predicate involving shared variables associated with each condition wait that is true if the thread should proceed. Spurious wakeups from the pthread_cond_timedwait() or pthread_cond_wait() functions may occur. Since the return from pthread_cond_timedwait() or pthread_cond_wait() does not imply anything about the value of this predicate, the predicate should be re-evaluated upon such return.

https://linux.die.net/man/3/pthread_cond_wait

Reply:
damn it. I didn't want to do this. But you're correct. Will update.